File: src/lib/ipc/testcases.ts
import { invoke } from '@tauri-apps/api/core';

export interface TestCase {
	id: string;
	name: string;
	description: string;
	category: string;
	tags: string;
	content: string;
	expected_message_type: string;
	expected_validation_result: string;
	created_at: string;
	updated_at: string;
}

export async function saveTestCase(
	testCase: Partial<TestCase> & {
		name: string;
		content: string;
	}
): Promise<TestCase> {
	return invoke('save_test_case', {
		id: testCase.id ?? null,
		name: testCase.name,
		description: testCase.description ?? '',
		category: testCase.category ?? 'general',
		tags: testCase.tags ?? '',
		content: testCase.content,
		expectedMessageType: testCase.expected_message_type ?? '',
		expectedValidationResult: testCase.expected_validation_result ?? 'valid',
	});
}

export async function getTestCases(category?: string): Promise<TestCase[]> {
	return invoke('get_test_cases', { category: category ?? null });
}

export async function deleteTestCase(id: string): Promise<void> {
	return invoke('delete_test_case', { id });
}

File: src-tauri/src/commands/test_cases.rs
use tauri::State;
use uuid::Uuid;

use crate::database::{Database, TestCase};
use crate::licensing::feature_gate;

#[tauri::command]
pub fn save_test_case(
    id: Option<String>,
    name: String,
    description: String,
    category: String,
    tags: String,
    content: String,
    expected_message_type: String,
    expected_validation_result: String,
    db: State<'_, Database>,
) -> Result<TestCase, String> {
    // Community cap: block only NEW saves beyond the limit. Existing cases
    // (even ones saved over the cap during a trial) stay editable/runnable.
    if let Some(max) = feature_gate::test_case_limit() {
        let existing = db.get_test_cases(None)?;
        let is_update = id
            .as_deref()
            .map(|i| existing.iter().any(|tc| tc.id == i))
            .unwrap_or(false);
        if !is_update && existing.len() >= max {
            feature_gate::require("test_cases_unlimited")?;
        }
    }

    let tc = TestCase {
        id: id.unwrap_or_else(|| Uuid::new_v4().to_string()),
        name, description, category, tags, content,
        expected_message_type, expected_validation_result,
        created_at: chrono::Utc::now().to_rfc3339(),
        updated_at: chrono::Utc::now().to_rfc3339(),
    };
    db.save_test_case(&tc)?;
    Ok(tc)
}

#[tauri::command]
pub fn get_test_cases(
    category: Option<String>,
    db: State<'_, Database>,
) -> Result<Vec<TestCase>, String> {
    db.get_test_cases(category.as_deref())
}

#[tauri::command]
pub fn delete_test_case(id: String, db: State<'_, Database>) -> Result<(), String> {
    db.delete_test_case(&id)
}
